### PR TITLE
damage model revision

### PR DESCRIPTION
**Pull request by shadisharba**
This PR improves the convergence of the damage model by implementing an accurate Jacobian and by changing the time discretisation to an implicit Euler scheme.

### CODE REVIEW

File: src/constitutive/mechanics/solid/small_strain_J2_plasticity_damage.cpp [hunk 1]
```diff
@@ -63,186 +63,228 @@ void small_strain_J2_plasticity_damage::update_internal_variables(double const t
...
-                                              material.lambda(),
-                                              strain - plastic_strain);
+        cauchy_stress = compute_cauchy_stress_scalar_damage(material.shear_modulus(),
+                                                            material.lambda(),
+                                                            strain - plastic_strain,
+                                                            scalar_damage);
 
-        matrix3 tau = deviatoric(cauchy_stress) / (1.0 - scalar_damage) - deviatoric(back_stress);
+        // Trial state
+        matrix3 tau = cauchy_stress - back_stress;
```
Comment (dbeurle) [inline src/constitutive/mechanics/solid/small_strain_J2_plasticity_damage.cpp]: Is it possible to change that from `tau` to something like `trial_state` or `trial_stress`?

**Review by dbeurle — CHANGES_REQUESTED** [review]
Looks nice :+1: We can talk about the `tau` naming but otherwise tight.  Thanks!

File: src/constitutive/mechanics/solid/small_strain_J2_plasticity_damage.cpp (hunk 1 above)
Comment (shadisharba) [reply src/constitutive/mechanics/solid/small_strain_J2_plasticity_damage.cpp]: I've just found what it should be called "relative stress"

Comment (shadisharba) [reply src/constitutive/mechanics/solid/small_strain_J2_plasticity_damage.cpp]: Done. I added one extra function to compute the energy release rate which makes it easier to introduce crack-closure effect later if needed.

**Review by dbeurle — APPROVED** [review]

**Pull request merged**